File: backend/src/igab/api/v1/schemas/simplefin.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import field_validator

from igab.api.v1.schemas.base import ApiModel
from igab.integrations.simplefin.limits import GLOBAL_DAILY_LIMIT


class SimpleFINSetupRequest(ApiModel):
    setup_token: str


class SimpleFINConfigResponse(ApiModel):
    """Whether this server can store bank credentials at all.

    The client cannot work this out — the encryption key is server-side env —
    so it asks before offering the setup form. All three fields are required:
    a path that forgets one must raise, not report a misconfigured server as
    ready and let the user spend a single-use token finding out.

    ``generate_key_command`` is served rather than written into the frontend so
    the recipe lives in exactly one place (``simplefin.encryption``), beside
    the check that decides whether a key is acceptable.
    """

    configured: bool
    problem: str | None
    generate_key_command: str


class SimpleFINUpdateRequest(ApiModel):
    sync_enabled: bool | None = None
    #: UTC hours (0-23) to sync at; an empty list is "never". Omitted means
    #: "leave the schedule alone" — the route drops None fields.
    sync_hours: list[int] | None = None

    @field_validator("sync_hours")
    @classmethod
    def _canonical_hours(cls, hours: list[int] | None) -> list[int] | None:
        """Sorted, deduplicated, in range, and within the daily budget.

        The cap is the connection's own rate limit rather than a number picked
        here: scheduling a 13th sync would only queue a request the provider
        refuses, and the error should say so at the moment it is set rather
        than silently at 3am.
        """
        if hours is None:
            return None
        if any(h < 0 or h > 23 for h in hours):
            raise ValueError("sync hours must be between 0 and 23")
        unique = sorted(set(hours))
        if len(unique) > GLOBAL_DAILY_LIMIT:
            raise ValueError(
                f"at most {GLOBAL_DAILY_LIMIT} syncs a day — that is this "
                "connection's daily limit with SimpleFIN"
            )
        return unique


class SimpleFINConnectionResponse(ApiModel):
    id: uuid.UUID
    user_id: uuid.UUID
    last_sync_at: datetime | None
    sync_enabled: bool
    sync_hours: list[int]
    global_requests_today: int
    account_requests_today: int
    last_sync_error: str | None
    last_sync_error_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LinkSimpleFINRequest(ApiModel):
    simplefin_account_id: str
    simplefin_account_name: str | None = None


class RefetchRequest(ApiModel):
    """Which connection to ask. An account does not record its connection —
    the link is by bank account id — and the modal already knows which one
    it listed the account from."""

    connection_id: uuid.UUID


class OrphanedLinkInfo(ApiModel):
    """An account whose bank link the feed no longer offers.

    Carries its own suggested replacement so the UI can offer a one-click
    relink rather than making the user match bank strings by eye.
    """

    account_id: uuid.UUID
    account_name: str
    stored_simplefin_id: str
    suggested_feed_id: str | None = None
    suggested_feed_name: str | None = None
    #: The bridge reported an institution needing re-authentication in the
    #: same response and offered no replacement: the account is unreachable,
    #: not gone, and the fix is at the bridge rather than a relink here.
    may_need_auth: bool = False


class BankErrorInfo(ApiModel):
    """One entry from the bridge's `errlist`. `con.auth` means an institution
    needs re-authenticating, and the bridge's guide is explicit that these
    must be shown to the user."""

    code: str
    message: str
    connection_id: str | None = None


class BalanceDriftInfo(ApiModel):
    """A reconciled account whose ledger disagrees with the bank after a sync.

    The bank's figure was always stored; what was missing was anyone saying
    so at the moment it changed. A gap here is the sync's own admission that
    something did not arrive.
    """

    account_id: uuid.UUID
    account_name: str
    bank_balance: Decimal
    ledger_cleared_balance: Decimal


class SyncResult(ApiModel):
    imported: int
    skipped: int
    #: Why each skipped row was skipped. A single count conflating "belonged
    #: to no linked account" with "already filed" is what made a nine-day
    #: outage read as a normal sync.
    skip_reasons: dict[str, int] = {}
    matched: int = 0
    #: Rows whose bank id was replaced wholesale and re-stamped onto the
    #: existing row, rather than imported beside it as a duplicate.
    adopted: int = 0
    review_queued: int = 0
    cleared: int = 0
    removed_pending: int = 0
    orphaned_links: list[OrphanedLinkInfo] = []
    bank_errors: list[BankErrorInfo] = []
    balance_drift: list[BalanceDriftInfo] = []
    #: Accounts whose first sync wrote a Starting Balance row to anchor the
    #: ledger to the bank's reported balance — the 90-day window cannot
    #: carry an older carried balance any other way.
    anchored: int = 0
    error: str | None = None
    global_used: int | None = None
    global_remaining: int | None = None
    account_used: int | None = None
    account_remaining: int | None = None


class ConnectionSyncOutcome(ApiModel):
    """What one connection did during a sync-all."""

    connection_id: uuid.UUID
    imported: int = 0
    skipped: int = 0
    adopted: int = 0
    error: str | None = None
    #: Carried per connection, not just in the totals: a broken link on one
    #: bank is the whole story of that connection's run.
    orphaned_links: list[OrphanedLinkInfo] = []
    bank_errors: list[BankErrorInfo] = []
    balance_drift: list[BalanceDriftInfo] = []


class SyncAllResult(ApiModel):
    """Every connection's sync, totalled.

    One failing connection does not stop the others, so the totals and the
    per-connection list are both needed: "imported 4" is not the whole story
    when a second bank was rate-limited.
    """

    imported: int
    skipped: int
    skip_reasons: dict[str, int] = {}
    matched: int = 0
    adopted: int = 0
    review_queued: int = 0
    cleared: int = 0
    removed_pending: int = 0
    anchored: int = 0
    connections: list[ConnectionSyncOutcome] = []


class RateLimitStatus(ApiModel):
    global_used: int
    global_remaining: int
    account_used: int
    account_remaining: int
    can_sync_global: bool
    can_sync_account: bool
    resets_at: str


class AccountSyncStatusResponse(ApiModel):
    account_id: uuid.UUID
    simplefin_account_id: str | None
    simplefin_sync_enabled: bool
    first_sync_complete: bool
    last_simplefin_sync_at: datetime | None
    simplefin_balance: Decimal | None

    model_config = {"from_attributes": True}


class TransactionMatchResponse(ApiModel):
    id: uuid.UUID
    synced_transaction_id: uuid.UUID
    manual_transaction_id: uuid.UUID
    confidence_score: Decimal
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class SyncRunAccountResponse(ApiModel):
    """What one run did to one account.

    `feed_txn_count == 0` on an account the run was told to sync is the
    signature of a bank link that no longer resolves, and `feed_newest_date`
    is what says an account has quietly stopped receiving anything.
    """

    account_id: uuid.UUID | None = None
    account_name: str | None = None
    simplefin_account_id: str | None = None
    feed_txn_count: int = 0
    feed_oldest_date: date | None = None
    feed_newest_date: date | None = None
    imported: int = 0
    adopted: int = 0
    reidentified: bool = False
    orphaned: bool = False
    bank_balance: Decimal | None = None
    ledger_cleared_balance: Decimal | None = None

    model_config = {"from_attributes": True}


class SyncRunResponse(ApiModel):
    id: uuid.UUID
    connection_id: uuid.UUID | None = None
    trigger: str
    status: str
    window_start: datetime | None = None
    window_end: datetime | None = None
    duration_ms: int | None = None
    error: str | None = None
    bank_errors: list[BankErrorInfo] = []
    orphaned_links: list[OrphanedLinkInfo] = []
    balance_drift: list[BalanceDriftInfo] = []
    feed_txn_count: int = 0
    imported: int = 0
    skipped: int = 0
    skip_reasons: dict[str, int] = {}
    matched: int = 0
    adopted: int = 0
    cleared: int = 0
    review_queued: int = 0
    removed_pending: int = 0
    anchored: int = 0
    #: Present when the run's writes can be taken back as a unit.
    change_batch_id: uuid.UUID | None = None
    undone_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class SyncRunUndoResult(ApiModel):
    """What "Undo this run" managed. `skipped` are changes left in place
    because the row was edited since — the person's later word wins."""

    undone: int
    skipped: int


class UnservedAccount(ApiModel):
    """An account the latest run was told to sync and the feed offered
    nothing for — no rows and no balance. The link is broken or the
    institution is unreachable; either way the account has stopped."""

    account_id: uuid.UUID
    account_name: str | None = None


class SyncRunDetailResponse(SyncRunResponse):
    accounts: list[SyncRunAccountResponse] = []


class SyncRunListResponse(ApiModel):
    runs: list[SyncRunResponse]
    total_count: int


class SyncHealthResponse(ApiModel):
    """Whether anything about bank sync needs attention right now.

    Read from the most recent run rather than the most recent *problem*: the
    question is "is this broken now", and a stale finding would answer it
    wrongly in both directions.
    """

    orphaned_links: list[OrphanedLinkInfo] = []
    needs_auth: list[BankErrorInfo] = []
    #: Re-checked against the ledger as it is now, not as the run left it,
    #: so deleting the duplicates clears the badge without another sync.
    balance_drift: list[BalanceDriftInfo] = []
    unserved: list[UnservedAccount] = []
    last_run_at: datetime | None = None

    @property
    def clean(self) -> bool:
        return not (self.orphaned_links or self.needs_auth or self.balance_drift or self.unserved)

File: frontend/src/api/simplefin.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'
import { invalidateAfterImport } from './invalidateAfterImport'
import type {
  SimpleFINConfig,
  SimpleFINConnection,
  SimpleFINRateLimitStatus,
  SyncResult,
  TransactionMatch,
} from '../types'
import { ROOT } from './queryKeys'

/**
 * Whether the server can run bank sync. Asked before the setup form is shown:
 * a SimpleFIN setup token is single-use, so a misconfigured server has to say
 * so before the user spends one on it.
 */
export function useSimpleFINConfig() {
  return useQuery({
    queryKey: [ROOT.simplefinConfig],
    queryFn: async () => {
      const { data } = await apiClient.get<SimpleFINConfig>('/simplefin/config')
      return data
    },
    staleTime: 60_000,
  })
}

export function useSimpleFINConnections() {
  return useQuery({
    queryKey: [ROOT.simplefinConnections],
    queryFn: async () => {
      const { data } = await apiClient.get<SimpleFINConnection[]>('/simplefin/connections')
      return data
    },
    staleTime: 30_000,
  })
}

export function useSimpleFINRateLimitStatus(connectionId: string | null) {
  return useQuery({
    queryKey: [ROOT.simplefinRateLimit, connectionId],
    queryFn: async () => {
      const { data } = await apiClient.get<SimpleFINRateLimitStatus>(
        `/simplefin/connections/${connectionId}/status`
      )
      return data
    },
    enabled: !!connectionId,
    staleTime: 0,
    refetchOnMount: 'always',
    refetchInterval: 30_000,
  })
}

export function useSetupSimpleFIN() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (setup_token: string) =>
      apiClient.post<SimpleFINConnection>('/simplefin/setup', { setup_token }).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.simplefinConnections] })
    },
  })
}

export function useUpdateSimpleFINConnection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      id,
      ...updates
    }: {
      id: string
      sync_enabled?: boolean
      /** Omit to leave the schedule alone; [] turns it off. */
      sync_hours?: number[]
    }) =>
      apiClient
        .put<SimpleFINConnection>(`/simplefin/connections/${id}`, updates)
        .then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.simplefinConnections] })
    },
  })
}

export function useDeleteSimpleFINConnection() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) => apiClient.delete(`/simplefin/connections/${id}`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.simplefinConnections] })
    },
  })
}

export function useSyncSimpleFIN(budgetId: string | null) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      connectionId,
      accountSimplefinId,
    }: {
      connectionId: string
      accountSimplefinId?: string
    }) =>
      apiClient
        .post<SyncResult>(
          `/simplefin/connections/${connectionId}/sync`,
          {},
          {
            params: {
              budget_id: budgetId,
              ...(accountSimplefinId ? { account_simplefin_id: accountSimplefinId } : {}),
            },
          }
        )
        .then((r) => r.data),
    onSuccess: () => {
      // A sync is an import: rows, payees it created, balances, budget math,
      // hygiene — the shared sweep covers what per-key lists kept missing
      // (new payees rendered "—" until their staleTime lapsed).
      invalidateAfterImport(qc, budgetId)
      qc.invalidateQueries({ queryKey: [ROOT.simplefinConnections] })
      qc.invalidateQueries({ queryKey: [ROOT.simplefinRateLimit] })
      qc.invalidateQueries({ queryKey: [ROOT.simplefinMatches] })
    },
  })
}

/** An account whose bank link the feed no longer offers. */
export interface OrphanedLink {
  account_id: string
  account_name: string
  stored_simplefin_id: string
  suggested_feed_id: string | null
  suggested_feed_name: string | null
}

/** One entry from the bridge's `errlist`. `con.auth` means re-authentication. */
export interface BankError {
  code: string
  message: string
  connection_id: string | null
}

export interface ConnectionSyncOutcome {
  connection_id: string
  imported: number
  skipped: number
  adopted: number
  error: string | null
  orphaned_links: OrphanedLink[]
  bank_errors: BankError[]
}

export interface SyncAllResult {
  imported: number
  skipped: number
  /** Why rows were skipped, keyed by `SkipReason`. See formatSyncSummary. */
  skip_reasons?: Record<string, number>
  matched: number
  adopted: number
  review_queued: number
  cleared: number
  removed_pending: number
  connections: ConnectionSyncOutcome[]
}

/**
 * Sync every connection, not just the first one.
 *
 * The Accounts page's "Sync All" posted to `connections[0]`, so a household
 * with two banks only ever synced one of them from it. The loop lives on the
 * server, where the rate limit and the connection list already are, and this
 * is what the Accounts page, the sidebar and the command palette all call.
 */
export function useSyncAllSimpleFIN(budgetId: string | null) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async () => {
      const { data } = await apiClient.post<SyncAllResult>(`/${budgetId}/simplefin/sync-all`)
      return data
    },
    onSuccess: () => {
      // A sync is an import — same sweep the single-connection sync uses.
      invalidateAfterImport(qc, budgetId)
      qc.invalidateQueries({ queryKey: [ROOT.simplefinConnections] })
      qc.invalidateQueries({ queryKey: [ROOT.simplefinRateLimit] })
      qc.invalidateQueries({ queryKey: [ROOT.simplefinMatches] })
    },
  })
}

/**
 * One sentence for what a sync did, so the three places that can start one
 * cannot describe the same run differently.
 *
 * A broken bank link leads, because it is the only outcome here the user can
 * act on and the only one that means an account has silently stopped
 * importing. This message used to read "Imported 0, skipped 586" for exactly
 * that situation — indistinguishable from a quiet morning with nothing new,
 * which is how it went unnoticed for nine days.
 *
 * Errors are counted rather than listed: with several connections the message
 * has to stay a line long, and the connection carries its own error for the
 * settings page to show in full.
 */
export function formatSyncSummary(result: SyncAllResult): string {
  const orphans = result.connections.flatMap((c) => c.orphaned_links ?? [])
  if (orphans.length > 0) {
    const names = orphans.map((o) => o.account_name)
    const who = names.length === 1 ? names[0] : `${names.length} accounts`
    return `${who} no longer match an account at the bank — relink to resume importing`
  }

  const needsAuth = result.connections
    .flatMap((c) => c.bank_errors ?? [])
    .filter((e) => e.code === 'con.auth')
  if (needsAuth.length > 0 && result.imported === 0) {
    const who = needsAuth.length === 1 ? 'A bank' : `${needsAuth.length} banks`
    return `${who} needs re-authenticating — see SimpleFIN in Settings`
  }

  const failed = result.connections.filter((c) => c.error).length
  const parts = [`Imported ${result.imported}`]
  if (result.adopted) parts.push(`re-linked ${result.adopted} existing`)
  if (result.matched) parts.push(`matched ${result.matched}`)
  if (result.cleared) parts.push(`cleared ${result.cleared}`)
  if (result.review_queued) parts.push(`${result.review_queued} need review`)
  // "Nothing new" is the honest reading when every skip was a row already
  // filed, and it is what the benign case actually means.
  parts.push(describeSkips(result))
  const summary = parts.filter(Boolean).join(', ')
  if (failed === 0) return summary
  const banks = failed === 1 ? '1 connection' : `${failed} connections`
  return `${summary} — ${banks} could not sync`
}

/** The `skipped` half of the summary, in the terms the count actually means. */
function describeSkips(result: SyncAllResult): string {
  if (result.skipped === 0) return ''
  const reasons = result.skip_reasons ?? {}
  const onlyAlreadyFiled =
    Object.keys(reasons).length > 0 && Object.keys(reasons).every((r) => r === 'already_posted')
  if (onlyAlreadyFiled) return `${result.skipped} already filed`
  const foreign = reasons.foreign_account ?? 0
  if (foreign === result.skipped && foreign > 0) {
    return `${foreign} belong to unlinked accounts`
  }
  return `skipped ${result.skipped}`
}

export function useLinkSimpleFINAccount(accountId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, name }: { id: string; name: string | null }) =>
      apiClient.post(`/accounts/${accountId}/link-simplefin`, {
        simplefin_account_id: id,
        simplefin_account_name: name,
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.accounts] })
    },
  })
}

export function useUnlinkSimpleFINAccount(accountId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () => apiClient.delete(`/accounts/${accountId}/link-simplefin`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.accounts] })
    },
  })
}

export function useUpdateAccountSimpleFINSettings(accountId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (simplefin_sync_enabled: boolean) =>
      apiClient.patch(`/accounts/${accountId}/simplefin-settings`, null, {
        params: { simplefin_sync_enabled },
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.accounts] })
    },
  })
}

export function useSimpleFINRemoteAccounts(connectionId: string | null) {
  return useQuery({
    queryKey: [ROOT.simplefinRemoteAccounts, connectionId],
    queryFn: async () => {
      const { data } = await apiClient.get<{ id: string; name: string }[]>(
        `/simplefin/connections/${connectionId}/accounts`
      )
      return data
    },
    enabled: !!connectionId,
    staleTime: 60_000,
  })
}

export function usePendingMatchesForAccount(accountId: string | null) {
  return useQuery({
    queryKey: [ROOT.pendingMatchesAccount, accountId],
    queryFn: async () => {
      const { data } = await apiClient.get<TransactionMatch[]>(
        `/accounts/${accountId}/pending-matches`
      )
      return data
    },
    enabled: !!accountId,
    staleTime: 15_000,
  })
}

export function usePendingMatches(budgetId: string | null) {
  return useQuery({
    queryKey: [ROOT.simplefinMatches, budgetId],
    queryFn: async () => {
      const { data } = await apiClient.get<TransactionMatch[]>('/simplefin/matches', {
        params: { budget_id: budgetId },
      })
      return data
    },
    enabled: !!budgetId,
    staleTime: 15_000,
  })
}

export function useAcceptMatch(accountId?: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (matchId: string) => apiClient.post(`/simplefin/matches/${matchId}/accept`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.simplefinMatches] })
      qc.invalidateQueries({ queryKey: [ROOT.pendingMatchesAccount, accountId] })
      qc.invalidateQueries({ queryKey: [ROOT.transactions] })
      qc.invalidateQueries({ queryKey: [ROOT.allTransactions] })
      // Accepting merges away the duplicate — cleared/working balances change
      qc.invalidateQueries({ queryKey: [ROOT.accounts] })
      qc.invalidateQueries({ queryKey: [ROOT.pendingReviewCount] })
      qc.invalidateQueries({ queryKey: [ROOT.pendingReviewCountAccount] })
    },
  })
}

export function useRejectMatch(accountId?: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (matchId: string) => apiClient.post(`/simplefin/matches/${matchId}/reject`),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.simplefinMatches] })
      qc.invalidateQueries({ queryKey: [ROOT.pendingMatchesAccount, accountId] })
    },
  })
}
